src/core/symbology/qgsinterpolatedlinerenderer.cpp

```diff
@@ -0,0 +1,676 @@
+#include "qgssymbollayerutils.h"
+
+void QgsInterpolatedLineRenderer::setInterpolatedWidth( const QgsInterpolatedLineWidth &strokeWidth )
+{
+  mStrokeWidth = strokeWidth;
+}
+
+void QgsInterpolatedLineRenderer::setInterpolatedColor( const QgsInterpolatedLineColor &strokeColoring )
+{
+  mStrokeColoring = strokeColoring;
+}
+
+void QgsInterpolatedLineRenderer::setWidthUnit( const QgsUnitTypes::RenderUnit &strokeWidthUnit )
+{
+  mStrokeWidthUnit = strokeWidthUnit;
+}
+
+void QgsInterpolatedLineRenderer::render( double value1, double value2, QgsPointXY point1, QgsPointXY point2, QgsRenderContext &context ) const
+{
+  QPainter *painter = context.painter();
+  painter->save();
+  if ( context.flags() & QgsRenderContext::Antialiasing )
+    painter->setRenderHint( QPainter::Antialiasing, true );
+  painter->setOpacity( mOpacity );
+
+  const QgsMapToPixel &mapToPixel = context.mapToPixel();
+
+  if ( value1 > value2 )
+  {
+    std::swap( value1, value2 );
+    std::swap( point1, point2 );
+  }
+
+  QPointF p1 = mapToPixel.transform( point1 ).toQPointF();
+  QPointF p2 = mapToPixel.transform( point2 ).toQPointF();
+  QPointF dir = p2 - p1;
+  double length = sqrt( pow( dir.x(), 2 ) + pow( dir.y(), 2 ) );
+  QPointF diru = dir / length;
+  QPointF orthu = QPointF( -diru.y(), diru.x() );
+
+  QList<double> breakValues;
+  QList<QColor> breakColors;
+  QList<QLinearGradient> gradients;
+
+  mStrokeColoring.graduatedColors( value1, value2, breakValues, breakColors, gradients );
+
+  if ( gradients.isEmpty() && !breakValues.empty() && !breakColors.isEmpty() ) //exact colors to render
+  {
+    Q_ASSERT( breakColors.count() == breakValues.count() );
+    for ( int i = 0; i < breakValues.count(); ++i )
+    {
+      double value = breakValues.at( i );
+      double width = context.convertToPainterUnits( mStrokeWidth.strokeWidth( value ), mStrokeWidthUnit );
+      QPen pen( breakColors.at( i ) );
+      pen.setWidthF( width );
+      pen.setCapStyle( Qt::PenCapStyle::RoundCap );
+      painter->setPen( pen );
+      QPointF point = p1 + dir * ( value - value1 ) / ( value2 - value1 );
+      painter->drawPoint( point );
+    }
+  }
+  else
+  {
+    double width1 = context.convertToPainterUnits( mStrokeWidth.strokeWidth( value1 ), mStrokeWidthUnit );
+    double width2 = context.convertToPainterUnits( mStrokeWidth.strokeWidth( value2 ), mStrokeWidthUnit ) ;
+
+    //Draw line cap
+    QRectF capBox1( p1.x() - width1 / 2, p1.y() - width1 / 2, width1, width1 );
+    QRectF capBox2( p2.x() - width2 / 2, p2.y() - width2 / 2, width2, width2 );
+    int startAngle;
+    startAngle = ( acos( -orthu.x() ) / M_PI ) * 180;
+    if ( orthu.y() < 0 )
+      startAngle = 360 - startAngle;
+
+    QBrush brush( Qt::SolidPattern );
+    QPen pen;
+
+    brush.setColor( mStrokeColoring.color( value1 ) );
+    painter->setBrush( brush );
+    pen.setBrush( brush );
+    painter->setPen( pen );
+
+    painter->drawPie( capBox1, ( startAngle - 1 ) * 16, 182 * 16 );
```

Comment:
For reference-- if you are successful and want to implement this for the standard linestring renderer, you can use QgsGeometry::taperedBuffer which works nicely with multi segment line strings.

Reply:
Thanks, I will look at!